### PR TITLE
testbench ipc abstraction and xt-run simplification

### PR DESCRIPTION
**Pull request by lgirdwood**
IPC abstraction to support IPC4 into the testbench and removal of testbench threading logic.

### CODE REVIEW

**Comment by lgirdwood** [comment]
@singalsu I've cleaned up some of the patches and put them on top of the IPC abstraction. Can you take a look. Thanks.

File: .gitignore [hunk 1]
```diff
@@ -45,6 +45,8 @@ boot_module
 tap-driver.sh
 test-driver
 *.trs
+tools/test/audio/reports/*
+tools/test/audio/zeros_in.raw
```
Comment (singalsu) [inline .gitignore:49]: Should fix the scripts/host-testbench.sh to delete this file after test.

**Review by singalsu — COMMENTED** [review]
Also some typos found by checkpatch

**Comment by lgirdwood** [comment]
@singalsu fwiw I'm not seeing this fuzzer build error locally
```
[ 90%] Building C object CMakeFiles/fuzz_ipc.dir/fuzz_ipc.c.o
/usr/local/bin/clang  -DRELATIVE_FILE=\"fuzz_ipc.c\" -I/src/sof/tools/oss-fuzz/build_oss_fuzz/sof_ep/install/include -O1 -fno-omit-frame-pointer -gline-tables-only -DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION -fsanitize=address -fsanitize-address-use-after-scope -fsanitize=fuzzer-no-link -g -O3 -Wall -Werror -Wmissing-prototypes -Wimplicit-fallthrough -DCONFIG_LIBRARY -imacros/src/sof/tools/oss-fuzz/build_oss_fuzz/sof_ep/build/library_autoconfig.h -MD -MT CMakeFiles/fuzz_ipc.dir/fuzz_ipc.c.o -MF CMakeFiles/fuzz_ipc.dir/fuzz_ipc.c.o.d -o CMakeFiles/fuzz_ipc.dir/fuzz_ipc.c.o -c /src/sof/tools/oss-fuzz/fuzz_ipc.c
[100%] Linking C executable /github/workspace/build-out/fuzz_ipc
/usr/local/bin/cmake -E cmake_link_script CMakeFiles/fuzz_ipc.dir/link.txt --verbose=1
/usr/local/bin/clang++ -O1 -fno-omit-frame-pointer -gline-tables-only -DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION -fsanitize=address -fsanitize-address-use-after-scope -fsanitize=fuzzer-no-link -fsanitize=fuzzer CMakeFiles/fuzz_ipc.dir/fuzz_ipc.c.o -o /github/workspace/build-out/fuzz_ipc  -Wl,-rpath,::::::::::::::::::::::::::::::::::::::::::::::::::::::::: -ldl -lm sof_ep/install/lib/libsof.a 
/usr/bin/ld: /usr/bin/ld: DWARF error: invalid or unhandled FORM value: 0x25
sof_ep/install/lib/libsof.a(dai.c.o): in function `dai_assign_group':
dai.c:(.text.dai_assign_group[dai_assign_group]+0x0): multiple definition of `dai_assign_group'; /usr/bin/ld: DWARF error: invalid or unhandled FORM value: 0x25
sof_ep/install/lib/libsof.a(dai-legacy.c.o):dai-legacy.c:(.text.dai_assign_group[dai_assign_group]+0x0): first defined here
/usr/bin/ld: /usr/bin/ld: DWARF error: invalid or unhandled FORM value: 0x25
sof_ep/install/lib/libsof.a(dai-legacy.c.o): in function `dai_comp_trigger_internal':
dai-legacy.c:(.text.dai_comp_trigger_internal[dai_comp_trigger_internal]+0xa9b): undefined reference to `platform_dai_wallclock'
clang-15: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: Leaving directory '/src/sof/tools/oss-fuzz/build_oss_fuzz'
make[2]: *** [CMakeFiles/fuzz_ipc.dir/build.make:98: /github/workspace/build-out/fuzz_ipc] Error 1
make[1]: Leaving directory '/src/sof/tools/oss-fuzz/build_oss_fuzz'
make[1]: *** [CMakeFiles/Makefile2:85: CMakeFiles/fuzz_ipc.dir/all] Error 2
make: *** [Makefile:136: all] Error 2
2023-03-24 22:27:59,180 - root - ERROR - Building fuzzers failed.
2023-03-24 22:27:59,181 - root - ERROR - Error building fuzzers for (commit: e20a2533c514359b73f7734e53c0f04483e4ad78, pr_ref: refs/pull/7331/merge).
```

File: .github/workflows/ipc_fuzzer.yml [hunk 2]
```diff
@@ -50,7 +50,7 @@ jobs:
   #
   # Then use a simple IPC3/IPC4 matrix like the one in zephyr.yml
   simple-IPC3-fuzz_sh:
-    runs-on: ubuntu-22.04
+    runs-on: ubuntu-22.10
```
Comment (cujomalainey) [inline .github/workflows/ipc_fuzzer.yml:53]: any reason to not use ubuntu-latest?

File: src/include/sof/audio/component.h [hunk 3]
```diff
@@ -699,8 +699,32 @@ static inline struct comp_dev *comp_alloc(const struct comp_driver *drv,
 #define comp_get_drvdata(c) \
 	(c->priv_data)
 
-#if defined UNIT_TEST || defined __ZEPHYR__
+#if defined UNIT_TEST || defined __ZEPHYR__  || CONFIG_LIBRARY_STATIC
 #define DECLARE_MODULE(init)
+
+/* declared modules */
+void sys_comp_asrc_init(void);
```
Comment (cujomalainey) [inline src/include/sof/audio/component.h:706]: random thought, any reason we are not putting a section in the memory map to dynamic registration in a library?

**Comment by cujomalainey** [comment]
> @singalsu fwiw I'm not seeing this fuzzer build error locally
> 
> ```
> [ 90%] Building C object CMakeFiles/fuzz_ipc.dir/fuzz_ipc.c.o
> /usr/local/bin/clang  -DRELATIVE_FILE=\"fuzz_ipc.c\" -I/src/sof/tools/oss-fuzz/build_oss_fuzz/sof_ep/install/include -O1 -fno-omit-frame-pointer -gline-tables-only -DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION -fsanitize=address -fsanitize-address-use-after-scope -fsanitize=fuzzer-no-link -g -O3 -Wall -Werror -Wmissing-prototypes -Wimplicit-fallthrough -DCONFIG_LIBRARY -imacros/src/sof/tools/oss-fuzz/build_oss_fuzz/sof_ep/build/library_autoconfig.h -MD -MT CMakeFiles/fuzz_ipc.dir/fuzz_ipc.c.o -MF CMakeFiles/fuzz_ipc.dir/fuzz_ipc.c.o.d -o CMakeFiles/fuzz_ipc.dir/fuzz_ipc.c.o -c /src/sof/tools/oss-fuzz/fuzz_ipc.c
> [100%] Linking C executable /github/workspace/build-out/fuzz_ipc
> /usr/local/bin/cmake -E cmake_link_script CMakeFiles/fuzz_ipc.dir/link.txt --verbose=1
> /usr/local/bin/clang++ -O1 -fno-omit-frame-pointer -gline-tables-only -DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION -fsanitize=address -fsanitize-address-use-after-scope -fsanitize=fuzzer-no-link -fsanitize=fuzzer CMakeFiles/fuzz_ipc.dir/fuzz_ipc.c.o -o /github/workspace/build-out/fuzz_ipc  -Wl,-rpath,::::::::::::::::::::::::::::::::::::::::::::::::::::::::: -ldl -lm sof_ep/install/lib/libsof.a 
> /usr/bin/ld: /usr/bin/ld: DWARF error: invalid or unhandled FORM value: 0x25
> sof_ep/install/lib/libsof.a(dai.c.o): in function `dai_assign_group':
> dai.c:(.text.dai_assign_group[dai_assign_group]+0x0): multiple definition of `dai_assign_group'; /usr/bin/ld: DWARF error: invalid or unhandled FORM value: 0x25
> sof_ep/install/lib/libsof.a(dai-legacy.c.o):dai-legacy.c:(.text.dai_assign_group[dai_assign_group]+0x0): first defined here
> /usr/bin/ld: /usr/bin/ld: DWARF error: invalid or unhandled FORM value: 0x25
> sof_ep/install/lib/libsof.a(dai-legacy.c.o): in function `dai_comp_trigger_internal':
> dai-legacy.c:(.text.dai_comp_trigger_internal[dai_comp_trigger_internal]+0xa9b): undefined reference to `platform_dai_wallclock'
> clang-15: error: linker command failed with exit code 1 (use -v to see invocation)
> make[2]: Leaving directory '/src/sof/tools/oss-fuzz/build_oss_fuzz'
> make[2]: *** [CMakeFiles/fuzz_ipc.dir/build.make:98: /github/workspace/build-out/fuzz_ipc] Error 1
> make[1]: Leaving directory '/src/sof/tools/oss-fuzz/build_oss_fuzz'
> make[1]: *** [CMakeFiles/Makefile2:85: CMakeFiles/fuzz_ipc.dir/all] Error 2
> make: *** [Makefile:136: all] Error 2
> 2023-03-24 22:27:59,180 - root - ERROR - Building fuzzers failed.
> 2023-03-24 22:27:59,181 - root - ERROR - Error building fuzzers for (commit: e20a2533c514359b73f7734e53c0f04483e4ad78, pr_ref: refs/pull/7331/merge).
> ```

how are you building? Are you using the oss-fuzz sdk or the scripts directly?

**Comment by lgirdwood** [comment]
> how are you building? Are you using the oss-fuzz sdk or the scripts directly?

Locally running the script, but I'm using 22.10 hence my workflow update test to use 22.10 i.e. to align clang and other tools to the same versions. This can be revivsted later, in the mean time I'll cherry-pick @singalsu clang fixes and drop the workflow update.

File: src/include/sof/audio/component.h (hunk 3 above)
Comment (lgirdwood) [reply src/include/sof/audio/component.h:706]: xt-testbench simulator usage is static, but see ```lrg/topic/plugin``` branch as that uses dynamic loading via ALSA plugin.

File: tools/testbench/topology.c [hunk 4]
```diff
@@ -684,21 +480,21 @@ static int load_fileread(struct tplg_context *ctx, int dir)
...
-	ret = tplg_load_filewrite(ctx, &filewrite);
+	ret = tb_new_filewrite(ctx, &filewrite);
 	if (ret < 0)
 		return ret;
 
-	if (tplg_create_controls(ctx, 0, NULL, 0) < 0) {
-		fprintf(stderr, "error: loading controls\n");
-		return -EINVAL;
-	}
+//	if (tplg_create_controls(ctx, 0, NULL, 0) < 0) {
```
Comment (singalsu) [inline tools/testbench/topology.c:410]: Something to clean up?

File: tools/testbench/common_test.c [hunk 5]
```diff
@@ -342,7 +275,7 @@ void debug_print(char *message)
 /* enable trace in testbench */
 void tb_enable_trace(bool enable)
 {
-	test_bench_trace = enable;
+	//test_bench_trace = enable;
```
Comment (singalsu) [inline tools/testbench/common_test.c:278]: Delete this?

File: src/ipc/ipc-helper.c [hunk 6]
```diff
@@ -116,6 +116,10 @@ int comp_verify_params(struct comp_dev *dev, uint32_t flag,
 		return -EINVAL;
 	}
 
+	tr_info(&dev->tctx, "params pre: rate %d channels %d format %d",
```
Comment (singalsu) [inline src/ipc/ipc-helper.c:119]: Typo in commit text "teh"

**Review by singalsu — APPROVED** [review]
All I tried to test with this version worked without problems. There's a lot of code changes but looked OK to me.

**Comment by cujomalainey** [comment]
> > how are you building? Are you using the oss-fuzz sdk or the scripts directly?
> 
> Locally running the script, but I'm using 22.10 hence my workflow update test to use 22.10 i.e. to align clang and other tools to the same versions. This can be revivsted later, in the mean time I'll cherry-pick @singalsu clang fixes and drop the workflow update.

you might be able to reproduce this if you pull the container directly and build in it https://google.github.io/oss-fuzz/advanced-topics/reproducing/#building-using-docker

**Comment by marc-hb** [comment]
Looking at the "unknown FORM" failure in https://github.com/thesofproject/sof/actions/runs/4536809159/jobs/7993943875?pr=7331 and googling for 5 minutes, it looks like the failure is caused by an inconsistency between the custom `/usr/local/bin/clang` manually installed in the oss-fuzz-base container and the `/usr/bin/ld` default provided Ubuntu 22.04 in that same oss-fuzz-base.

`/usr/local/bin/clang` 15.0 in gcr.io/oss-fuzz-base/base-builder:latest produces DWARF v5 but the standard GNU /usr/bin/ld in the Ubuntu 22.04 base of oss-fuzz-base does not support DWARF v5 yet?

BTW the "focal" lines in the build log seem to prove changing the `runs-on: ubuntu-xxx` change makes no difference between the work is done inside the other, oss-fuzz-base container.

However it's all hard to tell for sure because the oss-fuzz-base build logs are incredibly long - any idea why?


BTW do we really need oss-fuzz-base container and indirection still now that we have the newer and more direct fuzzing added by @andyross and commit c47221f22b80 ? No idea, just wondering.

**Comment by cujomalainey** [comment]
> However it's all hard to tell for sure because the oss-fuzz-base build logs are incredibly long - any idea why?

Its essentially doing a dockerbuild as well as a firmware build inside the clang container, so you are seeing things like an alsa build as well
 
> BTW do we really need oss-fuzz-base container and indirection still now that we have the newer and more direct fuzzing added by @andyross and commit [c47221f](https://github.com/thesofproject/sof/commit/c47221f22b8034bf13166254e59ac549b18f4d60) ? No idea, just wondering.

Andy's will replace this eventually. It is going to become the new ipc fuzzer, I just stole his focus for an internal project temporarily

File: tools/testbench/topology.c (hunk 4 above)
Comment (lgirdwood) [reply tools/testbench/topology.c]: fixed

File: tools/testbench/common_test.c (hunk 5 above)
Comment (lgirdwood) [reply tools/testbench/common_test.c:278]: fixed

File: src/audio/CMakeLists.txt [hunk 7]
```diff
@@ -6,7 +6,7 @@ elseif(CONFIG_IPC_MAJOR_4)
 	set(mixer_src mixin_mixout/mixin_mixout.c mixin_mixout/mixin_mixout_generic.c mixin_mixout/mixin_mixout_hifi3.c)
 endif()
 
-if(NOT CONFIG_LIBRARY)
+if((NOT CONFIG_LIBRARY) OR CONFIG_LIBRARY_STATIC)
```
Comment (kv2019i) [inline src/audio/CMakeLists.txt:9]: @lgirdwood Why we add these files to static library build?

File: src/audio/module_adapter/CMakeLists.txt [hunk 8]
```diff
@@ -1,10 +1,11 @@
...
 
-if(NOT CONFIG_LIBRARY)
-	if (CONFIG_IPC_MAJOR_4 OR CONFIG_CADENCE_CODEC OR (NOT CONFIG_COMP_LEGACY_INTERFACE) OR CONFIG_COMP_FIR)
+
+if (CONFIG_IPC_MAJOR_4 OR CONFIG_CADENCE_CODEC OR (NOT CONFIG_COMP_LEGACY_INTERFACE) OR CONFIG_COMP_FIR)
 	add_local_sources(sof module_adapter.c module/generic.c)
-	endif()
+endif()
 
+if((NOT CONFIG_LIBRARY) OR CONFIG_LIBRARY_STATIC)
```
Comment (kv2019i) [inline src/audio/module_adapter/CMakeLists.txt:8]: Here again, so I guess this is on purpose...?

**Review by kv2019i — APPROVED** [review]
Removal of ll-scheduler is a big change, but probably the xt-run model is more widely applicable (to uses beyond xt-run as well). A few minor notes, but nothing to block merge.

File: src/audio/module_adapter/CMakeLists.txt (hunk 8 above)
Comment (lgirdwood) [reply src/audio/module_adapter/CMakeLists.txt:8]: yep, we need to build in - however this does need to be updated to more meaningful CONFIG names as part of testbench/plugin work.

**Pull request merged**